[inductor] don't materialize the large sparse matrix in CE bwd

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129043

Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of Softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

The Fx graph snippets that construct this aforementioned sparse matrix looks like:
```
       full_default_3: "bf16[32768, 50257]" = torch.ops.aten.full.default([32768, 50257], 0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)
       scatter: "bf16[32768, 50257]" = torch.ops.aten.scatter.value(full_default_3, 1, where_2, -1.0);  full_default_3 = where_2 = None
```
Leveraging the following observations:
- the scatter is applied upon a all zero (or more generally a const tensor)
- the index tensor for the scatter has a single element on the scatter dimension. In this case it's the label tensor

allow us to lower this 'scatter_upon_const_tensor' pattern to a pointwise kernel that can be easily fused with downstream kernels:

```
    def inner_fn(idx):
        selector_idx = list(idx)
        selector_idx[dim] = 0  # can do this since the index tensor has a single element on the scatter dimension

        selector = selector_loader(selector_idx)
        return ops.where(
            selector == ops.index_expr(idx[dim], torch.int64),
            ops.constant(val, dtype),
            ops.constant(background_val, dtype),
        )
```

## Test result on microbenchmark

For the microbenchmark added as `test_cross_entropy_loss`, we improve latency from 47.340ms to 42.768ms, memory footprint from 10.524GB to 7.227GB on A100. (on H100, we improve latency from 27.54ms to 23.51ms, memory footprint from 10.574GB to 7.354GB).

The saving matches the back-of-envelope calculation. We avoid storing a BF16 tensor with shape [30K, 50K] which is about 3GB in size. On A100, avoid loading and storing such a tensor can roughly save 3GB x 2 / 1.5TBGS = 4ms

## Test result on llm.c

We also test this on llm.c and the saving is much larger especially for memory footprint. The reason is due to autotuning that allocates extra memory for benchmarking. (Check https://github.com/pytorch/pytorch/issues/129258 and https://github.com/pytorch/pytorch/pull/129399 for more details).

For llm.c PyTorch implementation on A100, we improve from
171K tokens/s , 33.6G peak memory usage to
180K tokens/s, 18.6G peak memory usage. (A **45%** saving of peak memory)

## Test on PyTorch 2.0 Dashboard

The optimization is quite general especially for transformers. We tested this on PyTorch2.0 dashboard. Here is the [result](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Mon%2C%2017%20Jun%202024%2018%3A07%3A51%20GMT&stopTime=Mon%2C%2024%20Jun%202024%2018%3A07%3A51%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/158/head&lCommit=c62c55e29c65497d495217b6574bb36b0c4da7d4&rBranch=main&rCommit=0d25f096c1beaf8749932a3d6083ad653405ed71).

TLDR, for Huggingface benchmark suite, we get **6%** geomean perf improvement and **10%** geomean memory footprint improvement.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @jansel @Chillee @eellison 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

shunting314 committed:
[inductor] don't materialize the large sparse matrix in CE bwd

[ghstack-poisoned]

shunting314 committed:
Update on "[wip][inductor] don't materialize the large sparse matrix in CE bwd"


Still a prototype and need more testing and code cleanup.

But for microbenchmark https://gist.github.com/shunting314/f0b293d328a32fe140b9d22b147b492d , we can see that we improve both perf and memory usage significantly. The percent of saving on real model may be smaller since CrossEntropyLoss is only part of the model. But the win should still be decent.

Before the change:
ms=27.541120529174805, peak_mem=10.574 GB

After the change:
ms=23.505352020263672, peak_mem=7.354 GB

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[wip][inductor] don't materialize the large sparse matrix in CE bwd"


Still a prototype and need more testing and code cleanup.

It is a bit tricky to do pattern matching in this case since the sparse matrix has 2 users:
- A sum node to handle ignore_index in CrossEntropyLoss
- A sub node to compute gradients of softmax input

It's not straightforward to match the two user nodes at the same time. But it turns out that there is a nice way to handle this with 2 separate patterns:
1. we register a graph pattern for the sum node. After matching and replacement, the use count for the sparse matrix will only be 1
2. we register a lowering pattern to handle the remaining usage of the sparse matrix.

For microbenchmark https://gist.github.com/shunting314/f0b293d328a32fe140b9d22b147b492d , we can see that we improve both perf and memory usage significantly. The percent of saving on real model may be smaller since CrossEntropyLoss is only part of the model. But the win should still be decent.

Before the change:
ms=27.541120529174805, peak_mem=10.574 GB

After the change:
ms=23.505352020263672, peak_mem=7.354 GB

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented:
What's the graph we're matching here?

jansel reviewed (changes requested):
Looks like a good improvement, but needs a test.

shunting314 commented:
> What's the graph we're matching here?

cc @Chillee : Here is the whole backward graph https://gist.github.com/shunting314/2082ee22613848b937ef992cc8717bd7 for the microbench I run.

And the matched part is:
```
        full_default_3: "bf16[32768, 50257]" = torch.ops.aten.full.default([32768, 50257], 0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)

        scatter: "bf16[32768, 50257]" = torch.ops.aten.scatter.value(full_default_3, 1, where_2, -1.0);  full_default_3 = where_2 = None

        mul: "bf16[32768, 50257]" = torch.ops.aten.mul.Tensor(scatter, where_3);  scatter = where_3 = None

        convert_element_type_6: "f32[32768, 50257]" = torch.ops.prims.convert_element_type.default(mul, torch.float32);  mul = None
        
        sum_4: "f32[32768, 1]" = torch.ops.aten.sum.dim_IntList(convert_element_type_6, [1], True)

        sub_2: "f32[32768, 50257]" = torch.ops.aten.sub.Tensor(convert_element_type_6, mul_1);  convert_element_type_6 = mul_1 = None
```

shunting314 committed:
Update on "[wip][inductor] don't materialize the large sparse matrix in CE bwd"


Still a prototype and need more testing and code cleanup.

Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

It is a bit tricky to do pattern matching in this case since the sparse matrix has 2 users:
- A sum node to handle ignore_index in CrossEntropyLoss
- A sub node to compute gradients of softmax input

It's not straightforward to match the two user nodes at the same time. But it turns out that there is a nice way to handle this with 2 separate patterns:
1. we register a graph pattern for the sum node. After matching and replacement, the use count for the sparse matrix will only be 1
2. we register a lowering pattern to handle the remaining usage of the sparse matrix.

For microbenchmark https://gist.github.com/shunting314/f0b293d328a32fe140b9d22b147b492d , we can see that we improve both perf and memory usage significantly. The percent of saving on real model may be smaller since CrossEntropyLoss is only part of the model. But the win should still be decent.

Before the change:
ms=27.541120529174805, peak_mem=10.574 GB

After the change:
ms=23.505352020263672, peak_mem=7.354 GB

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[wip][inductor] don't materialize the large sparse matrix in CE bwd"


Still a prototype and need more testing and code cleanup.

Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

It is a bit tricky to do pattern matching in this case since the sparse matrix has 2 users:
- A sum node to handle ignore_index in CrossEntropyLoss
- A sub node to compute gradients of softmax input

It's not straightforward to match the two user nodes at the same time. But it turns out that there is a nice way to handle this with 2 separate patterns:
1. we register a graph pattern for the sum node. After matching and replacement, the use count for the sparse matrix will only be 1
2. we register a lowering pattern to handle the remaining usage of the sparse matrix.

For microbenchmark https://gist.github.com/shunting314/f0b293d328a32fe140b9d22b147b492d , we can see that we improve both perf and memory usage significantly. The percent of saving on real model may be smaller since CrossEntropyLoss is only part of the model. But the win should still be decent.

Before the change:
ms=27.541120529174805, peak_mem=10.574 GB

After the change:
ms=23.505352020263672, peak_mem=7.354 GB

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented:
Horace gave me a nice suggestion offline to pattern match only the scatter+allzero part and let inductor figure out the rest itself. It works well. This makes code much simpler and more general. And now I don't need to duplicate the patterns to make it work for llm.c (which does not have the convert element type node).

For llm.c, I see larger gain then I expected. I was only expecting a saving of 6GB memory (a [30k, 50k] fp32 tensor). But the benchmark results shows the memory usage drops from 33556 MiB -> 18595 MiB.  I would like to dig more on this

Chillee commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -216,6 +217,54 @@ def is_valid_mm_plus_mm(match: Match):
+@register_lowering_pattern(
+    CallFunction(
+        aten.scatter,
+        CallFunction(aten.full, KeywordArg("shape"), 0.0, dtype=KeywordArg("dtype")),
```
I think we can support an arbitrary `val` here, right?

Chillee replied:
Below you just need to rewrite it to 

```
        return ops.where(selector == ops.index_expr(idx[1], torch.int64), val, matched_constant)
```

shunting314 replied:
Good point. There is no harm to make it more general. Will do that.

I'll rename the patter from `scatter_upon_allzero` to `scatter_upon_const_tensor` then

Chillee replied:
The 2 op pattern match - optimal :^)

Chillee commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -216,6 +217,54 @@ def is_valid_mm_plus_mm(match: Match):
+    CallFunction(
+        aten.scatter,
+        CallFunction(aten.full, KeywordArg("shape"), 0.0, dtype=KeywordArg("dtype")),
+        1,  # dim
```
Is there any reason we need to restrict the dim?

Chillee replied:
I think might as well make it more general if it's easy enough.

eellison reviewed (commented):
```
It is a bit tricky to do pattern matching in this case since the sparse matrix has 2 users:

A sum node to handle ignore_index in CrossEntropyLoss
A sub node to compute gradients of softmax input

```

See, other pattern like this:

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_2.py#L40-L42

We should also add a `register_replacement_for_lowering` that handles this..

shunting314 commented on torch/_inductor/fx_passes/post_grad.py:
```diff
@@ -216,6 +217,54 @@ def is_valid_mm_plus_mm(match: Match):
+def scatter_upon_allzero(match: Match, shape, dtype, selector, val):
+    from torch._inductor import lowering
+
+    if len(selector.get_size()) == 2:
```
lowering pattern only works in post-grad pass, right?

I'm worried that if we express the replacement in aten, the default lowering rule of scatter would make it not fusable to consumer nodes.

eellison replied:
```
def decomp(selector, val):
    x_dim = torch.arange(selector.shape[0])
    y_dim = torch.arange(selector.shape[1])
    return torch.where(selector[x_dim] == y_dim, val, 0)
```

does something along those work ? need to think abt it

shunting314 replied:
nope, selector is a 1D tensor here.

eellison replied:
I still think we can do this. For example:


```
import torch
torch.set_default_device("cuda")

# Example usage
M, N = 1024, 2048
x = torch.randint(0, N, (M,), dtype=torch.int64)

def f(x):
    y = torch.full([M, N], 3.14, dtype=torch.float)
    y.scatter_(1, x.unsqueeze(1), 2.718)
    return y

def pointwise_scatter(x):
    y = torch.full((M, N), 3.14, dtype=torch.float, device=x.device)
    range_tensor = torch.arange(N, device=x.device).view(1, -1)
    x_expanded = x.unsqueeze(1).expand(-1, N)
    condition = (x_expanded == range_tensor)
    return torch.where(condition, torch.full_like(y, 2.718), y)
    
torch.testing.assert_allclose(optimized_scatter(x), f(x))
```

shunting314 replied:
Nice to see this can be expressed as aten ops! I didn't realize that we can use `arange` like this to generate all indices .

But as I understand, moving the pattern to joint graph is generally good since the partitioner may figure out something smarter. But it's may also be just neutral in practice. Since we've already done thorough tests on this PR, should we get this version in and do another round of tests with the equivalent graph pattern you proposed in a followup PR?

eellison replied:
Sure, we can do as a follow up (maybe a `good first issue` / `internal ramp up task`)

shunting314 replied:
Here it is: https://github.com/pytorch/pytorch/issues/129449

shunting314 committed:
Update on "[wip][inductor] don't materialize the large sparse matrix in CE bwd"


Still a prototype and need more testing and code cleanup.

Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

It is a bit tricky to do pattern matching in this case since the sparse matrix has 2 users:
- A sum node to handle ignore_index in CrossEntropyLoss
- A sub node to compute gradients of softmax input

It's not straightforward to match the two user nodes at the same time. But it turns out that there is a nice way to handle this with 2 separate patterns:
1. we register a graph pattern for the sum node. After matching and replacement, the use count for the sparse matrix will only be 1
2. we register a lowering pattern to handle the remaining usage of the sparse matrix.

For microbenchmark https://gist.github.com/shunting314/f0b293d328a32fe140b9d22b147b492d , we can see that we improve both perf and memory usage significantly. The percent of saving on real model may be smaller since CrossEntropyLoss is only part of the model. But the win should still be decent.

Before the change:
ms=27.541120529174805, peak_mem=10.574 GB

After the change:
ms=23.505352020263672, peak_mem=7.354 GB

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[wip][inductor] don't materialize the large sparse matrix in CE bwd"


Still a prototype and need more testing and code cleanup.

Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

It is a bit tricky to do pattern matching in this case since the sparse matrix has 2 users:
- A sum node to handle ignore_index in CrossEntropyLoss
- A sub node to compute gradients of softmax input

It's not straightforward to match the two user nodes at the same time. But it turns out that there is a nice way to handle this with 2 separate patterns:
1. we register a graph pattern for the sum node. After matching and replacement, the use count for the sparse matrix will only be 1
2. we register a lowering pattern to handle the remaining usage of the sparse matrix.

For microbenchmark https://gist.github.com/shunting314/f0b293d328a32fe140b9d22b147b492d , we can see that we improve both perf and memory usage significantly. The percent of saving on real model may be smaller since CrossEntropyLoss is only part of the model. But the win should still be decent.

Before the change:
ms=27.541120529174805, peak_mem=10.574 GB

After the change:
ms=23.505352020263672, peak_mem=7.354 GB

(These numbers are from testing on H100)

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[wip][inductor] don't materialize the large sparse matrix in CE bwd"


Still a prototype and need more testing and code cleanup.

Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

It is a bit tricky to do pattern matching in this case since the sparse matrix has 2 users:
- A sum node to handle ignore_index in CrossEntropyLoss
- A sub node to compute gradients of softmax input

It's not straightforward to match the two user nodes at the same time. But it turns out that there is a nice way to handle this with 2 separate patterns:
1. we register a graph pattern for the sum node. After matching and replacement, the use count for the sparse matrix will only be 1
2. we register a lowering pattern to handle the remaining usage of the sparse matrix.

For microbenchmark https://gist.github.com/shunting314/f0b293d328a32fe140b9d22b147b492d , we can see that we improve both perf and memory usage significantly. The percent of saving on real model may be smaller since CrossEntropyLoss is only part of the model. But the win should still be decent.

Before the change:
ms=27.541120529174805, peak_mem=10.574 GB

After the change:
ms=23.505352020263672, peak_mem=7.354 GB

(These numbers are from testing on H100)

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[wip][inductor] don't materialize the large sparse matrix in CE bwd"


Still a prototype and need more testing and code cleanup.

Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

It is a bit tricky to do pattern matching in this case since the sparse matrix has 2 users:
- A sum node to handle ignore_index in CrossEntropyLoss
- A sub node to compute gradients of softmax input

It's not straightforward to match the two user nodes at the same time. But it turns out that there is a nice way to handle this with 2 separate patterns:
1. we register a graph pattern for the sum node. After matching and replacement, the use count for the sparse matrix will only be 1
2. we register a lowering pattern to handle the remaining usage of the sparse matrix.

For microbenchmark https://gist.github.com/shunting314/f0b293d328a32fe140b9d22b147b492d , we can see that we improve both perf and memory usage significantly. The percent of saving on real model may be smaller since CrossEntropyLoss is only part of the model. But the win should still be decent.

Before the change:
ms=27.541120529174805, peak_mem=10.574 GB

After the change:
ms=23.505352020263672, peak_mem=7.354 GB

(These numbers are from testing on H100)

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented on torch/_inductor/metrics.py:
```diff
@@ -46,6 +46,7 @@
 cpp_outer_loop_fused_inner_counts: List[int] = []
 
 num_comprehensive_padding = 0
+num_matches_for_scatter_upon_const_tensor = 0
```
I think this is kinda weird lol. Oddly specific thing to add a counter for imo.

Chillee commented on test/inductor/test_scatter_optimization.py:
```diff
@@ -0,0 +1,118 @@
+        metrics.reset()
+
+    def check_metric(self, val=1):
+        self.assertEqual(val, metrics.num_matches_for_scatter_upon_const_tensor)
```
Can we just compare metrics against the `num_bytes_accessed`?

shunting314 replied:
`num_bytes_accessed` does not work correctly for mutation. We have a StarDep injected to maintain write-after-write dependencies. That StarDep should not contribute any memory access but we count it when estimate number of memory access.

In spite of the issue mentioned, I think we can still use 'num_bytes_accessed' to verify if the optimization kicks in . When the optimization kicks in, we still access less memory even though the sparse matrix is saved to a buffer in the unit test.

shunting314 replied:
I've added assertion for `metrics.num_bytes_accessed` for majority of test cases.

But for `test_cross_entropy_loss` , it's non-trivial to estimate the amount of memory access manually. And I don't like putting the printed number in unit test directly either since that does not help with understanding.

I think we can probably keep both:
1. have assertion for num_bytes_accessed
2. have a specific metric tracking if the optimization is triggered. This is especially useful when the model is a bit complex. It's also useful in production models if we want to know if specific optimizations get triggered. (probably need improve tlparse to expose metrics)

WDYT?

Chillee replied:
@shunting314 Don't we already have some infra for checking when patterns get triggered? I think we should be able to reuse that.

shunting314 replied:
Do you mean something like this:
```
            self.assertEqual(counters["inductor"]["pattern_matcher_count"], count)
            self.assertEqual(counters["inductor"]["pattern_matcher_nodes"], nodes)
        
```

Chillee replied:
@shunting314 yeah.

Chillee commented on torch/_inductor/config.py:
```diff
@@ -401,6 +401,10 @@ def fx_graph_remote_cache_default():
 is_nightly_or_source = "dev" in torch.__version__ or "git" in torch.__version__
 developer_warnings = is_fbcode() or is_nightly_or_source
 
+optimize_scatter_upon_const_tensor = (
```
Add a short comment describing this pattern.

Chillee commented on test/inductor/test_scatter_optimization.py:
```diff
@@ -0,0 +1,118 @@
+DO_PERF_TEST = os.environ.get("DO_PERF_TEST") == "1"
+
+
+class TestScatterOpt(TestCase):
```
I would also add some tests where this pattern shouldn't trigger.

Chillee reviewed (changes requested):
probably not wip anymore :P

shunting314 committed:
Update on "[inductor] don't materialize the large sparse matrix in CE bwd"


Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of Softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

The Fx graph snippets that construct this aforementioned sparse matrix looks like:
```
       full_default_3: "bf16[32768, 50257]" = torch.ops.aten.full.default([32768, 50257], 0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)
       scatter: "bf16[32768, 50257]" = torch.ops.aten.scatter.value(full_default_3, 1, where_2, -1.0);  full_default_3 = where_2 = None
```
Leveraging the following observations:
- the scatter is applied upon a all zero (or more generally a const tensor)
- the index tensor for the scatter has a single element on the scatter dimension. In this case it's the label tensor

allow us to lower this 'scatter_upon_const_tensor' pattern to a pointwise kernel that can be easily fused with downstream kernels:

```
    def inner_fn(idx):
        selector_idx = list(idx)
        selector_idx[dim] = 0  # can do this since the index tensor has a single element on the scatter dimension

        selector = selector_loader(selector_idx)
        return ops.where(
            selector == ops.index_expr(idx[dim], torch.int64),
            ops.constant(val, dtype),
            ops.constant(background_val, dtype),
        )
```

## Test result on microbenchmark

For the microbenchmark added as `test_cross_entropy_loss`, we improve latency from 47.340ms to 42.768ms, memory footprint from 10.524GB to 7.227GB on A100. (on H100, we improve latency from 27.54ms to 23.51ms, memory footprint from 10.574GB to 7.354GB).

The saving matches the back-of-envelope calculation. We avoid storing a BF16 tensor with shape [30K, 50K] which is about 3GB in size. On A100, avoid loading and storing such a tensor can roughly save 3GB x 2 / 1.5TBGS = 4ms

## Test result on llm.c

We also test this on llm.c and the saving is much larger especially for memory footprint. The reason is due to autotuning that allocates extra memory for benchmarking. (Check https://github.com/pytorch/pytorch/issues/129258 and https://github.com/pytorch/pytorch/pull/129399 for more details).

For llm.c PyTorch implementation on A100, we improve from
171K tokens/s , 33.6G peak memory usage to
180K tokens/s, 18.6G peak memory usage. (A **45%** saving of peak memory)

## Test on PyTorch 2.0 Dashboard

The optimization is quite general especially for transformers. We tested this on PyTorch2.0 dashboard. Here is the [result](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Mon%2C%2017%20Jun%202024%2018%3A07%3A51%20GMT&stopTime=Mon%2C%2024%20Jun%202024%2018%3A07%3A51%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/158/head&lCommit=c62c55e29c65497d495217b6574bb36b0c4da7d4&rBranch=main&rCommit=0d25f096c1beaf8749932a3d6083ad653405ed71).

TLDR, for Huggingface benchmark suite, we get **6%** geomean perf improvement and **10%** geomean memory footprint improvement.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] don't materialize the large sparse matrix in CE bwd"


Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of Softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

The Fx graph snippets that construct this aforementioned sparse matrix looks like:
```
       full_default_3: "bf16[32768, 50257]" = torch.ops.aten.full.default([32768, 50257], 0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)
       scatter: "bf16[32768, 50257]" = torch.ops.aten.scatter.value(full_default_3, 1, where_2, -1.0);  full_default_3 = where_2 = None
```
Leveraging the following observations:
- the scatter is applied upon a all zero (or more generally a const tensor)
- the index tensor for the scatter has a single element on the scatter dimension. In this case it's the label tensor

allow us to lower this 'scatter_upon_const_tensor' pattern to a pointwise kernel that can be easily fused with downstream kernels:

```
    def inner_fn(idx):
        selector_idx = list(idx)
        selector_idx[dim] = 0  # can do this since the index tensor has a single element on the scatter dimension

        selector = selector_loader(selector_idx)
        return ops.where(
            selector == ops.index_expr(idx[dim], torch.int64),
            ops.constant(val, dtype),
            ops.constant(background_val, dtype),
        )
```

## Test result on microbenchmark

For the microbenchmark added as `test_cross_entropy_loss`, we improve latency from 47.340ms to 42.768ms, memory footprint from 10.524GB to 7.227GB on A100. (on H100, we improve latency from 27.54ms to 23.51ms, memory footprint from 10.574GB to 7.354GB).

The saving matches the back-of-envelope calculation. We avoid storing a BF16 tensor with shape [30K, 50K] which is about 3GB in size. On A100, avoid loading and storing such a tensor can roughly save 3GB x 2 / 1.5TBGS = 4ms

## Test result on llm.c

We also test this on llm.c and the saving is much larger especially for memory footprint. The reason is due to autotuning that allocates extra memory for benchmarking. (Check https://github.com/pytorch/pytorch/issues/129258 and https://github.com/pytorch/pytorch/pull/129399 for more details).

For llm.c PyTorch implementation on A100, we improve from
171K tokens/s , 33.6G peak memory usage to
180K tokens/s, 18.6G peak memory usage. (A **45%** saving of peak memory)

## Test on PyTorch 2.0 Dashboard

The optimization is quite general especially for transformers. We tested this on PyTorch2.0 dashboard. Here is the [result](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Mon%2C%2017%20Jun%202024%2018%3A07%3A51%20GMT&stopTime=Mon%2C%2024%20Jun%202024%2018%3A07%3A51%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/158/head&lCommit=c62c55e29c65497d495217b6574bb36b0c4da7d4&rBranch=main&rCommit=0d25f096c1beaf8749932a3d6083ad653405ed71).

TLDR, for Huggingface benchmark suite, we get **6%** geomean perf improvement and **10%** geomean memory footprint improvement.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] don't materialize the large sparse matrix in CE bwd"


Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of Softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

The Fx graph snippets that construct this aforementioned sparse matrix looks like:
```
       full_default_3: "bf16[32768, 50257]" = torch.ops.aten.full.default([32768, 50257], 0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)
       scatter: "bf16[32768, 50257]" = torch.ops.aten.scatter.value(full_default_3, 1, where_2, -1.0);  full_default_3 = where_2 = None
```
Leveraging the following observations:
- the scatter is applied upon a all zero (or more generally a const tensor)
- the index tensor for the scatter has a single element on the scatter dimension. In this case it's the label tensor

allow us to lower this 'scatter_upon_const_tensor' pattern to a pointwise kernel that can be easily fused with downstream kernels:

```
    def inner_fn(idx):
        selector_idx = list(idx)
        selector_idx[dim] = 0  # can do this since the index tensor has a single element on the scatter dimension

        selector = selector_loader(selector_idx)
        return ops.where(
            selector == ops.index_expr(idx[dim], torch.int64),
            ops.constant(val, dtype),
            ops.constant(background_val, dtype),
        )
```

## Test result on microbenchmark

For the microbenchmark added as `test_cross_entropy_loss`, we improve latency from 47.340ms to 42.768ms, memory footprint from 10.524GB to 7.227GB on A100. (on H100, we improve latency from 27.54ms to 23.51ms, memory footprint from 10.574GB to 7.354GB).

The saving matches the back-of-envelope calculation. We avoid storing a BF16 tensor with shape [30K, 50K] which is about 3GB in size. On A100, avoid loading and storing such a tensor can roughly save 3GB x 2 / 1.5TBGS = 4ms

## Test result on llm.c

We also test this on llm.c and the saving is much larger especially for memory footprint. The reason is due to autotuning that allocates extra memory for benchmarking. (Check https://github.com/pytorch/pytorch/issues/129258 and https://github.com/pytorch/pytorch/pull/129399 for more details).

For llm.c PyTorch implementation on A100, we improve from
171K tokens/s , 33.6G peak memory usage to
180K tokens/s, 18.6G peak memory usage. (A **45%** saving of peak memory)

## Test on PyTorch 2.0 Dashboard

The optimization is quite general especially for transformers. We tested this on PyTorch2.0 dashboard. Here is the [result](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Mon%2C%2017%20Jun%202024%2018%3A07%3A51%20GMT&stopTime=Mon%2C%2024%20Jun%202024%2018%3A07%3A51%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/158/head&lCommit=c62c55e29c65497d495217b6574bb36b0c4da7d4&rBranch=main&rCommit=0d25f096c1beaf8749932a3d6083ad653405ed71).

TLDR, for Huggingface benchmark suite, we get **6%** geomean perf improvement and **10%** geomean memory footprint improvement.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] don't materialize the large sparse matrix in CE bwd"


Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of Softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

The Fx graph snippets that construct this aforementioned sparse matrix looks like:
```
       full_default_3: "bf16[32768, 50257]" = torch.ops.aten.full.default([32768, 50257], 0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)
       scatter: "bf16[32768, 50257]" = torch.ops.aten.scatter.value(full_default_3, 1, where_2, -1.0);  full_default_3 = where_2 = None
```
Leveraging the following observations:
- the scatter is applied upon a all zero (or more generally a const tensor)
- the index tensor for the scatter has a single element on the scatter dimension. In this case it's the label tensor

allow us to lower this 'scatter_upon_const_tensor' pattern to a pointwise kernel that can be easily fused with downstream kernels:

```
    def inner_fn(idx):
        selector_idx = list(idx)
        selector_idx[dim] = 0  # can do this since the index tensor has a single element on the scatter dimension

        selector = selector_loader(selector_idx)
        return ops.where(
            selector == ops.index_expr(idx[dim], torch.int64),
            ops.constant(val, dtype),
            ops.constant(background_val, dtype),
        )
```

## Test result on microbenchmark

For the microbenchmark added as `test_cross_entropy_loss`, we improve latency from 47.340ms to 42.768ms, memory footprint from 10.524GB to 7.227GB on A100. (on H100, we improve latency from 27.54ms to 23.51ms, memory footprint from 10.574GB to 7.354GB).

The saving matches the back-of-envelope calculation. We avoid storing a BF16 tensor with shape [30K, 50K] which is about 3GB in size. On A100, avoid loading and storing such a tensor can roughly save 3GB x 2 / 1.5TBGS = 4ms

## Test result on llm.c

We also test this on llm.c and the saving is much larger especially for memory footprint. The reason is due to autotuning that allocates extra memory for benchmarking. (Check https://github.com/pytorch/pytorch/issues/129258 and https://github.com/pytorch/pytorch/pull/129399 for more details).

For llm.c PyTorch implementation on A100, we improve from
171K tokens/s , 33.6G peak memory usage to
180K tokens/s, 18.6G peak memory usage. (A **45%** saving of peak memory)

## Test on PyTorch 2.0 Dashboard

The optimization is quite general especially for transformers. We tested this on PyTorch2.0 dashboard. Here is the [result](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Mon%2C%2017%20Jun%202024%2018%3A07%3A51%20GMT&stopTime=Mon%2C%2024%20Jun%202024%2018%3A07%3A51%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/158/head&lCommit=c62c55e29c65497d495217b6574bb36b0c4da7d4&rBranch=main&rCommit=0d25f096c1beaf8749932a3d6083ad653405ed71).

TLDR, for Huggingface benchmark suite, we get **6%** geomean perf improvement and **10%** geomean memory footprint improvement.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] don't materialize the large sparse matrix in CE bwd"


Inductor currently materialize a large sparse matrix in the backward pass for CrossEntropyLoss and load that to compute gradients of Softmax input. If we could fuse the sparse matrix computation to the consumer sides, we gonna have both perf and memory usage wins.

The Fx graph snippets that construct this aforementioned sparse matrix looks like:
```
       full_default_3: "bf16[32768, 50257]" = torch.ops.aten.full.default([32768, 50257], 0, dtype = torch.bfloat16, layout = torch.strided, device = device(type='cuda', index=0), pin_memory = False)
       scatter: "bf16[32768, 50257]" = torch.ops.aten.scatter.value(full_default_3, 1, where_2, -1.0);  full_default_3 = where_2 = None
```
Leveraging the following observations:
- the scatter is applied upon a all zero (or more generally a const tensor)
- the index tensor for the scatter has a single element on the scatter dimension. In this case it's the label tensor

allow us to lower this 'scatter_upon_const_tensor' pattern to a pointwise kernel that can be easily fused with downstream kernels:

```
    def inner_fn(idx):
        selector_idx = list(idx)
        selector_idx[dim] = 0  # can do this since the index tensor has a single element on the scatter dimension

        selector = selector_loader(selector_idx)
        return ops.where(
            selector == ops.index_expr(idx[dim], torch.int64),
            ops.constant(val, dtype),
            ops.constant(background_val, dtype),
        )
```

## Test result on microbenchmark

For the microbenchmark added as `test_cross_entropy_loss`, we improve latency from 47.340ms to 42.768ms, memory footprint from 10.524GB to 7.227GB on A100. (on H100, we improve latency from 27.54ms to 23.51ms, memory footprint from 10.574GB to 7.354GB).

The saving matches the back-of-envelope calculation. We avoid storing a BF16 tensor with shape [30K, 50K] which is about 3GB in size. On A100, avoid loading and storing such a tensor can roughly save 3GB x 2 / 1.5TBGS = 4ms

## Test result on llm.c

We also test this on llm.c and the saving is much larger especially for memory footprint. The reason is due to autotuning that allocates extra memory for benchmarking. (Check https://github.com/pytorch/pytorch/issues/129258 and https://github.com/pytorch/pytorch/pull/129399 for more details).

For llm.c PyTorch implementation on A100, we improve from
171K tokens/s , 33.6G peak memory usage to
180K tokens/s, 18.6G peak memory usage. (A **45%** saving of peak memory)

## Test on PyTorch 2.0 Dashboard

The optimization is quite general especially for transformers. We tested this on PyTorch2.0 dashboard. Here is the [result](https://hud.pytorch.org/benchmark/compilers?dashboard=torchinductor&startTime=Mon%2C%2017%20Jun%202024%2018%3A07%3A51%20GMT&stopTime=Mon%2C%2024%20Jun%202024%2018%3A07%3A51%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/158/head&lCommit=c62c55e29c65497d495217b6574bb36b0c4da7d4&rBranch=main&rCommit=0d25f096c1beaf8749932a3d6083ad653405ed71).

TLDR, for Huggingface benchmark suite, we get **6%** geomean perf improvement and **10%** geomean memory footprint improvement.

cc jansel Chillee eellison 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

eellison commented:
Pointwise cat also does some examination of pointwise inputs' inner_fn_opcount. But could be good to annotate cats we know we are going to fuse through for partitioner
